Release v0.3.0 (#740)
kinvolk/lokomotive: CHANGELOG.md

## v0.3.0 - 2020-07-29

We're happy to announce the release of Lokomotive v0.3.0 (Coast Starlight).

This release packs new features and bugfixes. Some of the highlights are:

> * Kubernetes 1.18.6
> * AKS Kubernetes 1.16.10

Reviewer: This seems confusing. Maybe something like:

```suggestion
* Kubernetes 1.18.6
* For Lokomotive clusters running on top of AKS, Kubernetes 1.16.10 is now used
```

Author: ```suggestion
* Kubernetes 1.18.6
* For Lokomotive clusters running on top of AKS, Kubernetes 1.16.10 is installed.
```

How does that sound ?

Reviewer: Perfect! :)

* [Component updates](#component-updates)

### Changes in v0.3.0

#### Kubernetes updates

- Update Kubernetes to
[v1.18.6](https://github.com/kubernetes/kubernetes/blob/master/CHANGELOG/CHANGELOG-1.18.md#v1186)
([#726](https://github.com/kinvolk/lokomotive/pull/726)).

#### Platform updates

##### Packet
- Update default machine type from `t1.small.x86` to `c3.small.x86`, since
    `t1.small.x86` are EOL and no longer available in new Packet projects
    ([#612](https://github.com/kinvolk/lokomotive/pull/612)).

  If you haven't explicitly defined the controller_type and/or
  worker_pool.node_type configuration options, upgrading to this release will
  replace your controller and/or worker nodes with c3.small.x86 machines. To

>   avoid this, set these configuration options to the desired values.

Reviewer: Are you sure this will happen? I think the cluster will be broken, as terraform will delete the controllers (etcd cluster is broken) and create a new ones, and therefore loose all your data.

Am I missing something?

Author: Yes @rata If the machine type isn't explicity defined, it will amount to a creation of new nodes entirely.
In order to mitigate that existing clusters), the call is to explicitly state in the lokocfg file.

I copy pasted the above paragraph from the `Release Notes` in your PR #612

Reviewer: Oh, didn't know it comes from there, thanks!

I still think the release notes on that PR are probably wrong. I think the cluster will be destroyed and all etcd data will be lost if you haven't explicitly defined the controller types.

And for workers it will create downtime too.

Author: Yes @rata that's what I said will happen. It will be a new cluster.

Reviewer 2: We should add a bold warning here, that all cluster data will be lost if user upgrades without specifying it.

Author: added a **WARNING** note.

Reviewer 2: Hm, I would rephrase that, to mention, that if not done, you can lose the entire controlplane of your cluster. I'd explain why the type changed, from what to what and include a snippet which should be added if someone do not define machine type at the moment.

Author: The comment already mentions why the type changed and what changed. I've also added a config snippet.

```
- Update default machine type from `t1.small.x86` to `c3.small.x86`, since
    `t1.small.x86` are EOL and no longer available in new Packet projects
    ([#612](https://github.com/kinvolk/lokomotive/pull/612)).

  **WARNING**: If you haven't explicitly defined the controller_type and/or
  worker_pool.node_type configuration options, upgrading to this release will
  replace your controller and/or worker nodes with c3.small.x86 machines thereby
  loosing all your cluster data. To avoid this, set these configuration options
  to the desired values.

  Make sure that the below variables are explicity defined in your cluster
  configuration. This only applies to machine type `t1.small.x86`.

  ```hcl
  cluster "packet" {
    .
    .
    controller_type = "t1.small.x86"
    .
    .
    worker_pool "pool-name" {
      .
      node_type = "t1.small.x86"
      .
    }
  }
  ```
```

##### AKS

- Update Kubernetes version to 1.16.10
    ([#712](https://github.com/kinvolk/lokomotive/pull/712)).

#### Component updates

- openebs: update to 1.11.0 ([#673](https://github.com/kinvolk/lokomotive/pull/673)).

- calico: update to
  [v3.15.0](https://docs.projectcalico.org/release-notes/#v3150)
  ([#652](https://github.com/kinvolk/lokomotive/pull/652)).

#### UX

- prometheus-operator: Organize prometheus related variables under a
  `prometheus` block in the configuration
  ([#710](https://github.com/kinvolk/lokomotive/pull/710)).

- contour: Remove `ingress_hosts` from contour configuration
    ([#635](https://github.com/kinvolk/lokomotive/pull/635)).

#### Features

- Add `enable_toolbox` variable to [rook-ceph
component](https://github.com/kinvolk/lokomotive/blob/v0.3.0/docs/configuration-reference/components/rook-ceph)
([#649](https://github.com/kinvolk/lokomotive/pull/649)). This allows Rook Ceph
toolbox to manage and configure Ceph.

- Use ingress to expose Prometheus and `external_labels` for federated clusters

>     ([#710](https://github.com/kinvolk/lokomotive/pull/710)).

Author: added new changes as separate commit, once finalized will squash change long commits into one.

 references](https://github.com/kinvolk/lokomotive/blob/v0.3.0/docs/configuration-reference)
 ([#651](https://github.com/kinvolk/lokomotive/pull/651)).

- Update contour configuration reference for usage with AWS
   ([#674](https://github.com/kinvolk/lokomotive/pull/674)).

- Add documentation related to the usage of `clc_snippets` for Packet and AWS
    ([#657](https://github.com/kinvolk/lokomotive/pull/657)).

- Improve documentation on using remote
    [backends](https://github.com/kinvolk/lokomotive/blob/v0.3.0/docs/configuration-reference/backend/s3.md)
    ([#670](https://github.com/kinvolk/lokomotive/pull/670)).

- How to guide for setting up monitoring on Lokomotive
    ([#480](https://github.com/kinvolk/lokomotive/pull/480)).

- Add `codespell` section in [development
    documentation](https://github.com/kinvolk/lokomotive/blob/v0.3.0/docs/development/README.md)
    ([#700](https://github.com/kinvolk/lokomotive/pull/700)).

#### Bugfixes

- Remove contour ingress workaround(due to an [upstream issue](https://github.com/projectcontour/contour/issues/403))
  for ExternalDNS ([#635](https://github.com/kinvolk/lokomotive/pull/635)).

#### Development

> - Must use consistent build tags for all platforms for CI

Author: removed

    ([#600](https://github.com/kinvolk/lokomotive/pull/600))
    & ([#676](https://github.com/kinvolk/lokomotive/pull/676)).

> - Disable marking releases as `pre-releases` ([#678](https://github.com/kinvolk/lokomotive/pull/678)).

Reviewer: What? Why would a user care?

> - Disable marking releases as `pre-releases` ([#678](https://github.com/kinvolk/lokomotive/pull/678)).

Author: my thoughts exactly ? Why a user would care ?

> - Disable marking releases as `pre-releases` ([#678](https://github.com/kinvolk/lokomotive/pull/678)).

Reviewer 2: According to https://docs.github.com/en/github/administering-a-repository/managing-releases-in-a-repository:
>To notify users that the release is not ready for production and may be unstable, select This is a pre-release.

Previously, all releases have been marked as `pre-release`, which "meant" that they were not ready for production use. Now we do that anymore, which may indicate, that they are actually ready.

However, our motivation is a change in GitHub UI, which if you only have pre-releases, won't show any releases on the main repository page, which is not desired.

I know it's probably a small detail, but in my opinion it can be interpreted as important change.

Alternatively, maybe we don't need to mention that at all, as it has been changed for all releases we have.

> - Disable marking releases as `pre-releases` ([#678](https://github.com/kinvolk/lokomotive/pull/678)).

Author: If thats the case, then I'd lean on not including it in the release changelog. what do others think ?

> - Disable marking releases as `pre-releases` ([#678](https://github.com/kinvolk/lokomotive/pull/678)).

Author: removed.

> - Show release values in `--verbose` mode only

Author: updated.

    ([#627](https://github.com/kinvolk/lokomotive/pull/627)).

> - Change NewClientset to create Kubernetes Clientset from kubeconfig file

Author: Yes, its not that interesting , but thats why its in the Development section.

> - Change NewClientset to create Kubernetes Clientset from kubeconfig file

Reviewer 2: I think we should remove it and wait until we can add to the changelog, that asset directory is no longer needed to check cluster health. This is just a tiny and intermediate step towards that.

> - Change NewClientset to create Kubernetes Clientset from kubeconfig file

Author: removed.

content, instead of from file path
([#628](https://github.com/kinvolk/lokomotive/pull/628)).

> - Remove provider aliases from go templates

Author: Yes, its not that interesting , but thats why its in the Development section.

what others think @iaguis ? I'll remove them if its not needed.

> - Remove provider aliases from go templates

Reviewer 2: For correctness:
```suggestion
- Remove Terraform provider aliases from platforms code
```

    ([#617](https://github.com/kinvolk/lokomotive/pull/617)).

#### Miscellaneous

- Following flatcar-linux/Flatcar#123, Flatcar 2513.1.0 for ARM contains the dig
    binary so the workaround is no longer needed
    ([#703](https://github.com/kinvolk/lokomotive/pull/703)).

> - Make lint target fail upon invalid git configuration locally. ([#488](https://github.com/kinvolk/lokomotive/pull/488)).

Author: removed

- Improve error message for `wait-for-dns` output
    ([#735](https://github.com/kinvolk/lokomotive/pull/735)).

- Add `codespell` target to Makefile and `all` target ([#661](https://github.com/kinvolk/lokomotive/pull/661)).

### Upgrading from v0.2.1

#### Configuration syntax changes

There have been some minor changes to the configurations of following components:
* contour

> * rometheus-operator.

Reviewer: ```suggestion
* prometheus-operator.
```

Please make sure new the configuration structure is in place before the upgrade.

##### Contour component

Optional `ingress_hosts` variable is now removed.

old:

```hcl
component "contour" {
  .
  .
  ingress_hosts = ["*.example.lokomotive-k8s.net"]
}
```

new:

```hcl
component "contour" {
  .
  .
}

```

##### Prometheus-operator component

> * Variable `external_url` is now removed.

Reviewer 2: Hm, this is actually moved to `ingress.host` field. I think we should mention that to the user. And the valid value changed: now `http://` needs to be removed.

Author: ```
Variable `external_url` is now removed and now configured under
    `prometheus.ingress.host`. Remove the http scheme when configuring.
```

    }
    .
    .
  }
  .
  .
}
```

Check out the new syntax in the [Prometheus Operator configuration
reference](https://github.com/kinvolk/lokomotive/blob/v0.2.0/docs/configuration-reference/components/prometheus-operator.md)
for details.

##### Upgrade steps

Go to your cluster's directory and run the following command.

```
lokoctl cluster apply
```

The update process typically takes about 10 minutes.
After the update, running `lokoctl health` should result in an output similar to the following.

```
Node                     Ready    Reason          Message

lokomotive-controller-0  True     KubeletReady    kubelet is posting ready status
lokomotive-1-worker-0    True     KubeletReady    kubelet is posting ready status
lokomotive-1-worker-1    True     KubeletReady    kubelet is posting ready status
lokomotive-1-worker-2    True     KubeletReady    kubelet is posting ready status
Name      Status    Message              Error

etcd-0    True      {"health":"true"}
```

##### Post upgrade steps

##### Openebs

>It is highly recommended to schedule a downtime for the application using the
>OpenEBS PV while performing this upgrade. Also, make sure you have taken a

> >backup of the data before starting the below upgrade procedure.

Reviewer: Why the quote? For style/format or comes from openebs docs? If it's the latter, I'd reference that

Author: Would changing to a **NOTE** from a quote help ?

Openebs docs have upgrade procedures for upgrading the control plane components, jiva storage engine and cstore storage engine. I think pointing to the openebs docs would cause more confusion.

Reviewer: I didn't meant pasting the docs, just saying that if this is a quote from openebs docs, add a link to where they come from. Not paste the content, just a link

Author: Added the link to the openebs documentation.

> >backup of the data before starting the below upgrade procedure.

Reviewer 2: Hm, that may raise the questions about the backup strategy :smile:

  avoid this, set these configuration options to the desired values.

##### AKS

- Update Kubernetes version to 1.16.10
    ([#712](https://github.com/kinvolk/lokomotive/pull/712)).

#### Component updates

- openebs: update to 1.11.0 ([#673](https://github.com/kinvolk/lokomotive/pull/673)).

- calico: update to
  [v3.15.0](https://docs.projectcalico.org/release-notes/#v3150)
  ([#652](https://github.com/kinvolk/lokomotive/pull/652)).

#### UX

- prometheus-operator: Organize prometheus related variables under a
  `prometheus` block in the configuration
  ([#710](https://github.com/kinvolk/lokomotive/pull/710)).

- Use `prometheus.ingress.host` to expose Prometheus instead of
    `prometheus_external_url`
    ([#710](https://github.com/kinvolk/lokomotive/pull/710)).

- contour: Remove `ingress_hosts` from contour configuration
    ([#635](https://github.com/kinvolk/lokomotive/pull/635)).

#### Features

- Add `enable_toolbox` variable to [rook-ceph
component](https://github.com/kinvolk/lokomotive/blob/v0.3.0/docs/configuration-reference/components/rook-ceph)
([#649](https://github.com/kinvolk/lokomotive/pull/649)). This allows managing
and configuring Ceph using toolbox pod.

> - Add `external_labels` for federated clusters. This helps to identify identify

Reviewer 2: i think we should mention that this is a prometheus feature.

Author: ```suggestion
Add Prometheus feature `external_labels` for federated clusters to Prometheus
    operator component. This helps to identify identify metrics queried from
    different clusters.
```

   ([#674](https://github.com/kinvolk/lokomotive/pull/674)).

- Add documentation related to the usage of `clc_snippets` for Packet and AWS
    ([#657](https://github.com/kinvolk/lokomotive/pull/657)).

- Improve documentation on using remote
    [backends](https://github.com/kinvolk/lokomotive/blob/v0.3.0/docs/configuration-reference/backend/s3.md)
    ([#670](https://github.com/kinvolk/lokomotive/pull/670)).

- How to guide for setting up monitoring on Lokomotive
    ([#480](https://github.com/kinvolk/lokomotive/pull/480)).

- Add `codespell` section in [development
    documentation](https://github.com/kinvolk/lokomotive/blob/v0.3.0/docs/development/README.md)
    ([#700](https://github.com/kinvolk/lokomotive/pull/700)).

- Include a demo GIF in the readme
    ([#636](https://github.com/kinvolk/lokomotive/pull/636)).

#### Bugfixes

- Remove contour ingress workaround (due to an [upstream issue](https://github.com/projectcontour/contour/issues/403))
  for ExternalDNS ([#635](https://github.com/kinvolk/lokomotive/pull/635)).

#### Development

- Disable marking releases as `pre-releases` ([#678](https://github.com/kinvolk/lokomotive/pull/678)).

> - Do not show release values in terraform output

Reviewer 2: As it might be confused with project release.
```suggestion
- Do not show Helm release values in terraform output
```

    binary so the workaround is no longer needed
    ([#703](https://github.com/kinvolk/lokomotive/pull/703)).

- Improve error message for `wait-for-dns` output
    ([#735](https://github.com/kinvolk/lokomotive/pull/735)).

- Add `codespell` to enable spell check on all PRs ([#661](https://github.com/kinvolk/lokomotive/pull/661)).

### Upgrading from v0.2.1

#### Configuration syntax changes

There have been some minor changes to the configurations of following components:
* contour
* prometheus-operator.

Please make sure new the configuration structure is in place before the upgrade.

##### Contour component

Optional `ingress_hosts` variable is now removed.

old:

```hcl
component "contour" {
  .
  .
  ingress_hosts = ["*.example.lokomotive-k8s.net"]
}
```

new:

```hcl
component "contour" {
  .
  .
}

```

##### Prometheus-operator component

* Prometheus specific variables are now under a `prometheus` block.
* Variable `external_url` is now removed and now configured under
    `prometheus.ingress.host`. Remove the http scheme when configuring.
* A new optional `prometheus.ingress` sub-block is introduced to expose

> Prometheus over ingress.

Reviewer 2: Those 2 items should be swapped, to keep it in the logical order, which is:
1. introduction of new block
2. removal of old field (as it has been moved to newly added block).

old:

```hcl
component "prometheus-operator" {
  .
  .
  prometheus_metrics_retention = "14d"

>   prometheus_external_url      = "https://api.example.com/prometheus"

Reviewer 2: We should use the same value for old and new config version, so user can easily map them, as there is no new values introduced here.
```suggestion
  prometheus_external_url      = "https://prometheus.example.lokomotive-k8s.net"
```

  prometheus_storage_size      = "50GiB"
  prometheus_node_selector = {
    "kubernetes.io/hostname" = "worker3"
  }
  .
  .
}
```

new:

```hcl
component "prometheus-operator" {
  .
  .
  prometheus {
    metrics_retention = "14d"
    storage_size      = "50GiB"
    node_selector = {
      "kubernetes.io/hostname" = "worker3"
    }

    ingress {
      host                       = "prometheus.example.lokomotive-k8s.net"
      class                      = "contour"
      certmanager_cluster_issuer = "letsencrypt-production"
    }
    .
    .
  }
  .
  .
}
```

Check out the new syntax in the [Prometheus Operator configuration

> reference](https://github.com/kinvolk/lokomotive/blob/v0.2.0/docs/configuration-reference/components/prometheus-operator.md)

Reviewer 2: ```suggestion
reference](https://github.com/kinvolk/lokomotive/blob/v0.3.0/docs/configuration-reference/components/prometheus-operator.md)
```

lokomotive-controller-0  True     KubeletReady    kubelet is posting ready status
lokomotive-1-worker-0    True     KubeletReady    kubelet is posting ready status
lokomotive-1-worker-1    True     KubeletReady    kubelet is posting ready status
lokomotive-1-worker-2    True     KubeletReady    kubelet is posting ready status
Name      Status    Message              Error

etcd-0    True      {"health":"true"}
```

##### Post upgrade steps

##### Openebs

OpenEBS control plane components and data plane components work independently.
Even after the OpenEBS Control Plane components have been upgraded to 1.11.0,
the Storage Pools and Volumes (both jiva and cStor) will continue to work with
older versions.

>Upgrade functionality is still under active development. It is highly
>recommended to schedule a downtime for the application using the OpenEBS PV
>while performing this upgrade. Also, make sure you have taken a backup of the
>data before starting the below upgrade procedure. - [Openebs
documentation](https://github.com/openebs/openebs/blob/master/k8s/upgrades/README.md#step-3-upgrade-the-openebs-pools-and-volumes)

###### Upgrade cStor Pools

> * Extract the SPC name using `kubectl get spc`:

Reviewer 2: We should perhaps open an issue to automate that.

Author: yes, I've already added to the todo list.

Author: https://github.com/kinvolk/lokomotive/issues/744

> ## v0.3.0 - 2020-07-30

Reviewer: 31 :)

Reviewer 2: Probably on 08-03 as @ipochi is not around today :smile:

Author: Today :)

Author: done. will need you approvals again.

    ([#617](https://github.com/kinvolk/lokomotive/pull/617)).

#### Miscellaneous

- Following flatcar-linux/Flatcar#123, Flatcar 2513.1.0 for ARM contains the dig
    binary so the workaround is no longer needed
    ([#703](https://github.com/kinvolk/lokomotive/pull/703)).

- Improve error message for `wait-for-dns` output
    ([#735](https://github.com/kinvolk/lokomotive/pull/735)).

- Add `codespell` to enable spell check on all PRs ([#661](https://github.com/kinvolk/lokomotive/pull/661)).

### Upgrading from v0.2.1

#### Configuration syntax changes

There have been some minor changes to the configurations of following components:
* contour
* prometheus-operator.

Please make sure new the configuration structure is in place before the upgrade.

##### Contour component

Optional `ingress_hosts` variable is now removed.

old:

```hcl
component "contour" {
  .
  .
  ingress_hosts = ["*.example.lokomotive-k8s.net"]
}
```

new:

```hcl
component "contour" {
  .
  .
}

```

##### Prometheus-operator component

* Prometheus specific variables are now under a `prometheus` block.
* A new optional `prometheus.ingress` sub-block is introduced to expose
Prometheus over ingress.

> * Variable `external_url` is now removed and now configured under

Reviewer 2: As `variable` is:
```hcl
variable "foo" {}
```

```suggestion
* Attribute `external_url` is now removed and now configured under
```

Attribute, because this is how HCL refers to it: https://github.com/hashicorp/hcl/tree/hcl2. (BTW Terraform is inconsistent and calls it `arguments` :facepalm: https://www.terraform.io/docs/configuration/syntax.html)

>     `prometheus.ingress.host`. Remove the http scheme when configuring.

Reviewer 2: ```suggestion
    `prometheus.ingress.host`. Remove URL scheme (e.g. `https://`) and URI (e.g. `/prometheus`) when configuring. URI is no longer supported and protocol is always HTTPS.
```

old:

```hcl
component "prometheus-operator" {
  .
  .
  prometheus_metrics_retention = "14d"

>   prometheus_external_url      = "https://prometheus.example.lokomotive-k8s.net""

Reviewer 2: ```suggestion
  prometheus_external_url      = "https://prometheus.example.lokomotive-k8s.net"
```

  prometheus_storage_size      = "50GiB"
  prometheus_node_selector = {
    "kubernetes.io/hostname" = "worker3"
  }
  .
  .
}
```

new:

```hcl
component "prometheus-operator" {
  .
  .
  prometheus {
    metrics_retention = "14d"
    storage_size      = "50GiB"
    node_selector = {
      "kubernetes.io/hostname" = "worker3"
    }

    ingress {
      host                       = "prometheus.example.lokomotive-k8s.net"
      class                      = "contour"

>       certmanager_cluster_issuer = "letsencrypt-production"

Reviewer 2: Are those parameters both required?

Author: not really, not required.
removing.

#### Misc

> * Bootkube Docker images are now pulled using Docker protocol, as quay.io plans to deprecate pulling images using ACI ([#656](https://github.com/kinvolk/lokomotive/pull/656)).

Reviewer 2: This should go into separate commit (and thanks for fixing it!).
